pkg/server/server.go

```diff
@@ -297,23 +297,30 @@ func NewServerWithOpts(opts ...OpenFGAServiceV1Option) (*Server, error) {
 		opt(s)
 	}
 
-	s.checkOptions = []graph.LocalCheckerOption{
+	cycleDetectionCheckResolver := graph.NewCycleDetectionCheckResolver()
+	s.checkResolver = cycleDetectionCheckResolver
+
+	localChecker := graph.NewLocalChecker(
 		graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
-		graph.WithMaxConcurrentReads(s.maxConcurrentReadsForCheck),
-		graph.WithSingleflightResolver(),
-	}
+	)
+
+	cycleDetectionCheckResolver.SetDelegate(localChecker)
+	localChecker.SetDelegate(cycleDetectionCheckResolver)
 
 	if s.checkQueryCacheEnabled {
 		s.logger.Info("Check query cache is enabled and may lead to stale query results up to the configured query cache TTL",
 			zap.Duration("CheckQueryCacheTTL", s.checkQueryCacheTTL),
 			zap.Uint32("CheckQueryCacheLimit", s.checkQueryCacheLimit))
-		s.checkCache = ccache.New(
-			ccache.Configure[*graph.CachedResolveCheckResponse]().MaxSize(int64(s.checkQueryCacheLimit)),
-		)
-		s.checkOptions = append(s.checkOptions, graph.WithCachedResolver(
-			graph.WithExistingCache(s.checkCache),
+
+		cachedCheckResolver := graph.NewCachedCheckResolver(
+			graph.WithMaxCacheSize(int64(s.checkQueryCacheLimit)),
+			graph.WithLogger(s.logger),
 			graph.WithCacheTTL(s.checkQueryCacheTTL),
-		))
+		)
+		s.cachedCheckResolverCloser = cachedCheckResolver.Close
+
+		cachedCheckResolver.SetDelegate(localChecker)
+		cycleDetectionCheckResolver.SetDelegate(cachedCheckResolver)
```

Comment:
Why are we not using `NewLayeredCheckResolver` here, which appears to be made just for this use case?